internal/cluster/certificates.go

```diff
@@ -0,0 +1,371 @@
+	// Look up each certificate as a secret and populate the certificate/key
+	for _, certificate := range c {
+		secret := &corev1.Secret{}
+		key := client.ObjectKey{
+			Name:      certificate.SecretName(cluster.Name),
+			Namespace: cluster.Namespace,
+		}
+		if err := ctrlclient.Get(ctx, key, secret); err != nil {
+			if apierrors.IsNotFound(err) {
+				kp, err := generateCACert()
+				if err != nil {
+					return err
+				}
+				certificate.KeyPair = kp
+				certificate.Generated = true
+				continue
+			}
+			return errors.WithStack(err)
+		}
+		// If a user has a badly formatted secret it will prevent the cluster from working.
+		kp, err := secretToKeyPair(secret)
+		if err != nil {
+			return err
+		}
+		certificate.KeyPair = kp
+	}
+	return nil
+}
+
+// GenerateCertificates will generate any certificates that do not have KeyPair data.
+func (c Certificates) GenerateCertificates() error {
+	for _, certificate := range c {
+		if certificate.KeyPair == nil {
+			kp, err := generateCACert()
+			if err != nil {
+				return err
+			}
+			certificate.KeyPair = kp
+			certificate.Generated = true
+		}
+	}
+	return nil
+}
+
+// SaveGeneratedCertificates will save any certificates that have been generated as Kubernetes secrets.
+func (c Certificates) SaveGeneratedCertificates(ctx context.Context, ctrlclient client.Client, cluster *clusterv1.Cluster, config *bootstrapv1.KubeadmConfig) error {
+	for _, certificate := range c {
+		if !certificate.Generated {
+			continue
+		}
+		secret := certificate.AsSecret(cluster, config)
+		if err := ctrlclient.Create(ctx, secret); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+// GetOrCreateCertificates is a convenience function that wraps cluster bootstrap certificate behavior.
+func (c Certificates) GetOrCreateCertificates(ctx context.Context, ctrlclient client.Client, cluster *clusterv1.Cluster, config *bootstrapv1.KubeadmConfig) error {
+	// Get the certificates that exist
+	if err := c.GetCertificates(ctx, ctrlclient, cluster); err != nil {
+		return err
+	}
+
+	// Create the certificates that don't exist
+	if err := c.GenerateCertificates(); err != nil {
+		return err
+	}
+
+	// Save any certificates that have been generated
+	if err := c.SaveGeneratedCertificates(ctx, ctrlclient, cluster, config); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Certificate represents a single certificate CA.
+type Certificate struct {
+	Generated bool
+	Name      string
+	*certs.KeyPair
+}
+
+// SecretName is the expected name of the Kubernetes secret the certificate is saved in.
+func (c *Certificate) SecretName(clustername string) string {
+	return fmt.Sprintf("%s-%s", clustername, c.Name)
+}
+
+// Hashes hash all the certificates stored in a CA certificate.
+func (c *Certificate) Hashes() ([]string, error) {
+	certificates, err := cert.ParseCertsPEM(c.Cert)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse Cluster CA Certificate")
```

Comment:
It looks like this is only used to hash the cluster ca, but we probably should have a less specific error message here, and we can wrapf where it's called to indicate that it's for the cluster ca.